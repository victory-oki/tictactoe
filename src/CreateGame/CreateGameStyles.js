import styled from "styled-components";
import { white, green } from '../theme/colors';
export const ContainerStyled = styled.div`
  max-width: 112rem;
  margin: 0rem auto;

  .heading{
    color: ${white[100]};
    font-size: 2.4rem;
    text-align: center;
  }

  .form{
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    width: 100%;
    & >*{
      text-align: center;
    }
    p, .link{
      width: 100%;
      font-size: 1.6rem;
      text-align: center;
      color: ${white[100]};
    }
    .link{
      text-decoration: none;
      color: ${green[100]};
      transition: all 0.2s ease-in-out;
      &:hover{
        text-decoration: underline;
      }
    }
  }

  .p-inputgroup-addon{
    background: transparent;
    color: ${white[100]};
  }
  .p-inputtext{
    font-size: 1.6rem;
    background: transparent;
    color: ${white[100]};
    &::placeholder, &:enabled:hover{
      color: ${white[100]};
    }
  }

  .p-inputgroup, .p-button{
    width: 30rem;
    font-size: 1.6rem;
    margin-top: 2rem;
  }
  .p-inputgroup{
    height: 5rem;
  }

`
