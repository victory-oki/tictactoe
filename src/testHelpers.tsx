
import React from "react";
import TestRenderer from 'react-test-renderer';
import * as ShallowRenderer from 'react-test-renderer/shallow';

import { BrowserRouter as Router } from 'react-router-dom';
import { store } from "./store";
import { Provider } from "react-redux";

export const testSnapshot = (component: JSX.Element, shallow = false) => {
    const tree = shallow ?
      ShallowRenderer.createRenderer().render(component) :
      TestRenderer.create(component).toJSON()
    ;
    expect(tree).toMatchSnapshot();
};


export const wrapContext = (ui: JSX.Element) => (
    <Router>
        <Provider store={store}>
            {ui}
        </Provider>
    </Router>
)
