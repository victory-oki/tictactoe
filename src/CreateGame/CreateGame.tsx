import React, { useEffect, useState } from 'react';
import { ContainerStyled } from './CreateGameStyles';
import { InputText } from 'primereact/inputtext';
import { Button } from 'primereact/button';
import { useSocket } from '../socketContext';
import * as GameUtils from '../gameutils';
import { EVENTS, IjoinObj } from '../gameModel';
import { Link, NavigateFunction, useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { setPlayerName } from '../Game/gameSlice';
import { FormikErrors, FormikProps, useFormik } from 'formik';
import { classNames } from 'primereact/utils';

function CreateGame() {
    const [username, setUsername] = useState('');
    const socket = useSocket();
    const navigate: NavigateFunction = useNavigate();
    const dispatch = useDispatch();
    interface IformValues {name: string};
    const formik:FormikProps<IformValues> = useFormik<IformValues>({
        initialValues: {
            name: '',
        },
        validate: (data) => {
            let errors:FormikErrors<IformValues> = {};
            if (!data.name) {
                errors.name = 'Name is required.';
            }
            return errors;
        },
        onSubmit: (data) => {
            setUsername(data.name);
            handleCreateGame(data.name);
            formik.resetForm();
        }
    });
    useEffect(() => {
        setupSocketListeners();
        return () => {
            destroyListeners();
        }
    }, [username]);

    const setupSocketListeners = () => {
        socket.on(EVENTS.CREATE_SUCCESS, ({roomId}) => {
            dispatch(setPlayerName(username));
            navigate(`/lobby/${roomId}`, { replace: false });
        });
    }
    const destroyListeners = () => {
        socket.off(EVENTS.CREATE_SUCCESS)
    }

    const handleCreateGame = (username:string): void => {
        const joinObject: IjoinObj = { actionType: 'CREATE', username };
        socket.emit(EVENTS.JOIN, joinObject);
    }

    const isFormFieldValid = (name:keyof IformValues) => !!(formik?.touched[name] && formik?.errors[name]);
    const getFormErrorMessage = (name:keyof IformValues) => {
        return isFormFieldValid(name) && <small className="p-error">{formik.errors[name]}</small>;
    };

    return (
        <ContainerStyled>
            <h1 className="heading">Create Game</h1>
            <form className="form" onSubmit={formik.handleSubmit}>
                <div className="p-inputgroup">
                    <span className={classNames({ 'p-invalid': isFormFieldValid('name'), 'p-inputgroup-addon': true })}>
                        <i className="pi pi-user"></i>
                    </span>
                    <InputText placeholder="Username"id="name" name="name" value={formik.values.name} onChange={formik.handleChange} autoFocus className={classNames({ 'p-invalid': isFormFieldValid('name') })}/>
                </div>
                {getFormErrorMessage('name')}
                input
                <Button label="Create Game" className="p-button-success" />
                <p>or</p>
                <Link to={'/join'} className='link'>Join Existing game</Link>
            </form>
        </ContainerStyled>
    )
}

export default CreateGame
