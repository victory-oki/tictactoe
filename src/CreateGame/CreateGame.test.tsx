import React from "react";
import { render, screen } from "@testing-library/react";
import CreateGame from "./CreateGame";
import {wrapContext, testSnapshot} from '../testHelpers';
describe('Create Game', () => {
    const wrappedComponent: JSX.Element = wrapContext(<CreateGame />);
    test('renders Create  Game Component', async () => {
        render(
            wrappedComponent
        );
        const textbox = await screen.findAllByRole('textbox');
        expect(textbox).toHaveLength(1);
        const button = await screen.findAllByRole('button');
        expect(button).toHaveLength(1);
        const link = await screen.findAllByRole('link');
        expect(link).toHaveLength(1);
    });

    it('Renders Correctly', ()=>{
        testSnapshot(wrappedComponent)
    })
})

