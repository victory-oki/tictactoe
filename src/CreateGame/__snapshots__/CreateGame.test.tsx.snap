// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Game Renders Correctly 1`] = `
<div
  className="sc-bczRLJ cunqcc"
>
  <h1
    className="heading"
  >
    Create Game
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="p-inputgroup"
    >
      <span
        className="p-inputgroup-addon"
      >
        <i
          className="pi pi-user"
        />
      </span>
      <input
        autoFocus={true}
        className="p-inputtext p-component"
        id="name"
        name="name"
        onChange={[Function]}
        onInput={[Function]}
        onKeyPress={[Function]}
        onPaste={[Function]}
        placeholder="Username"
        value=""
      />
    </div>
    input
    <button
      aria-label="Create Game"
      className="p-button p-component p-button-success"
      disabled={false}
    >
      <span
        className="p-button-label p-c"
      >
        Create Game
      </span>
    </button>
    <p>
      or
    </p>
    <a
      className="link"
      href="/join"
      onClick={[Function]}
    >
      Join Existing game
    </a>
  </form>
</div>
`;
